lib/openhab/dsl/debouncer.rb

```diff
@@ -0,0 +1,241 @@
+# frozen_string_literal: true
+
+module OpenHAB
+  module DSL
+    #
+    # Provides the feature for debouncing calls to a given block.
+    # It is used by {DSL::Rules::BuilderDSL#debounce debounce rule guard} and
+    # {OpenHAB::DSL.debounce #debounce DSL method}
+    #
+    # The debouncer can filter events and only allow the events on the leading or trailing edge
+    # of the given interval. Its behavior can be customized through settings passed to its
+    # {initialize constructor}.
+    #
+    # The following timing diagram illustrates the incoming triggers and the actual executions
+    # that are allowed by {debounce} filter.
+    #
+    # - Each character position represents an instant of time,
+    #   and the space inbetween is the unit of time
+    # - The `T` line shows individual trigger events represented by numbers/letters, and
+    # - The `E` line shows the actual executions of the corresponding trigger that are
+    #   allowed by the debounce guard.
+    #
+    # ```java
+    # Leading edge trigger, interval: 5
+    # T: A.B...C...D..EFG.HI......JKLMNOPQRST....U.....
+    # E: A.....C......E....I......J....O....T....U.....
+    #
+    # Trailing edge trigger, interval: 5
+    # T: A.B...C...D..EFG.HI......JKLMNOPQRST....U.....
+    # E: |....B|....D.|....I......|....O....T....U.....
+    # Note: "|" indicates the start of the debounce period
+    # ```
+    #
+    class Debouncer
+      # @return [Range,nil] The range of accepted debounce period, or nil if debouncing is disabled.
+      attr_reader :interval
+
+      # @return [Duration, nil] The minimum idle time to stop debouncing.
+      attr_reader :idle_time
+
+      #
+      # Constructor to create a debouncer object.
+      #
+      # The constructor sets the options and behaviour of the debouncer when the {#call}
+      # method is called.
+      #
+      # Terminology:
+      # - `calls` are invocations of the {#call} method, i.e. the events that need to be throttled / debounced.
+      # - `executions` are the actual code executions of the given block. Executions usually occurs
+      #   less frequently than the call to the debounce method.
+      #
+      # @param [Duration,Range,nil] for
+      #   - {Duration}: The minimum interval between executions. The debouncer will not execute
+      #     the given block more often than this.
+      #   - {Range}: A range of {Duration}s for the minimum to maximum interval between executions.
+      #     The maximum interval specified in the range takes precedence over the `idle_time` parameter.
```

Comment:
this is confusing. at first reading I thought that meant "if you provide a range, the range.end is the same as the idle_time parameter, but I don't think that's true"

Reply:
I would like this to be as easy to understand as possible, so I'd really appreciate your feedback. I've updated this in the upcoming commit.